### ISSUE TITLE
bulkUpdate triggers rectifyAll

### ISSUE THREAD

**Issue by bajtos** [strongloop/loopback#1893]
The current implementation of `bulkUpdate` triggers expensive `rectifyAll` operation whenever a model instance was updated or deleted.

TL;DR: we need to improve [getIdFromWhereByModelId()](https://github.com/strongloop/loopback/blob/3b66fc1c8839f21d25c8af05d6073c83378666ae/lib/persisted-model.js#L1477-L1490) to detect filters that can match either one record with a known id, or no records at all.

**"UPDATE" operations**

To ensure we don't accidentally override changes made while the replication is in progress, we update the record only if all fields have the expected values, see [lib/persisted-model.js#L1245](https://github.com/strongloop/loopback/blob/3b66fc1c8839f21d25c8af05d6073c83378666ae/lib/persisted-model.js#L1245):

```js
Model.updateAll(current.toObject(), data, function(err, result) {
 // ...
});
```

This triggers "after save" hook with a query containing lots of properties in the "where" part, e.g.

```json
{
  "where": {
    "id":"278d7610-a4bb-11e5-afb5-b3fd49f730c6",
    "title": "buy eggs",
    "completed": false,
    "created": 1450355943409
  }
}
```

A filter like this is not recognised by our heuristics detecting whether a single model instance is affected, see [getIdFromWhereByModelId](https://github.com/strongloop/loopback/blob/3b66fc1c8839f21d25c8af05d6073c83378666ae/lib/persisted-model.js#L1477-L1490):

```js
    var whereKeys = Object.keys(where);
    if (whereKeys.length != 1) return undefined;

    var idName = Model.getIdName();
    if (whereKeys[0] !== idName) return undefined;
    // etc.
```

**"DELETE" operations**

The situation is similar as above, we delete the record only if all properties have expected values, see [lib/persisted-model.js#L1327](https://github.com/strongloop/loopback/blob/3b66fc1c8839f21d25c8af05d6073c83378666ae/lib/persisted-model.js#L1327):

```js
Model.deleteAll(current.toObject(), function(err, result) {
  // ...
});
```

This triggers "after save" hook with a query containing lots of properties in the "where" part, e.g.

```json
{
  "id":"278d7610-a4bb-11e5-afb5-b3fd49f730c6",
  "title": "buy eggs",
  "completed": false,
  "created": 1450355943409
}
```

Such query is not recognised as affecting a single model instance only, and thus a full scan (rectifyAll) is triggered.

Related: https://github.com/strongloop/loopback/pull/1894 and https://github.com/strongloop-internal/scrum-cs/issues/397

**Comment by Amir-61** [comment]
Connect to https://github.com/strongloop/loopback/pull/1899

**Comment by bajtos** [comment]
Fixed by #1899 and released as `loopback@2.26.2`.
